### PR TITLE
Add execution model diagnostic and functional tests

### PR DESCRIPTION
**Pull request by jvepsalainen-nv**
## Summary

Adds 53 regression tests for the Slang program execution model — compute
dispatch, groupshared memory and barriers, atomics, wave divergence and
reconvergence, pipeline stages (vertex/fragment/geometry/hull/domain/mesh
/ray-tracing), varying-parameter legalization, and fragment derivatives.
Test-only PR; no compiler changes.

Files: 16 in `tests/diagnostics/execution-model/` and 37 in
`tests/language-feature/execution-model/`.

## `tests/diagnostics/execution-model/` (16 tests)

Pins diagnostics on invalid execution-model input.

**Compute / `numthreads` / `WaveSize`**
- `non-positive-numthreads` — zero dimension
- `numthreads-negative-value` — negative dimension
- `spec-constant-numthreads` — spec constant as dimension on non-SPIRV
- `invalid-wave-size` — `[WaveSize(3)]` (must be power of 2 in [4, 128])

**System-value typing & direction**
- `invalid-dispatch-thread-id-type` — `float3` and `uint4` on `SV_DispatchThreadID`
- `sv-invalid-direction` — `inout` on a system-value parameter → E30702
- `sv-output-direction-mismatch` — `out` on an input-only SV (`SV_DispatchThreadID`); pins the OutModifier branch in `validateSystemValueSemantic`
- `stage-dependency-incompatible` — compute outputting `SV_Depth`; fragment inputting `SV_GroupID` → E30702 (two cases)

**Stage / capability**
- `unknown-stage-name-attribute` — `[shader("invalid")]` → unknown capability
- `capability-stage-incompatible` — callable entry calls a fragment-only function; pins the `isIncompatibleWith` set-algebra path in `slang-capability.cpp`
- `stage-incompatible-in-param` — invalid `in` parameter on callable shader → E39018 (pins both occurrences emitted today)
- `stage-incompatible-out-param` — invalid `out` parameter on intersection shader → E39017 + E41018

**Mesh shader outputs**
- `mesh-output-not-array` — non-array mesh output
- `mesh-output-array-no-size` — mesh output array missing size
- `mesh-output-not-out` — mesh output without `out`

**Ray tracing**
- `ray-payload-invalid-stage` — invalid stage in `read(...) / write(...)` payload qualifier

## `tests/language-feature/execution-model/` (37 tests)

Functional and emission coverage for runtime execution-model behavior.

**Compute dispatch & system values** (3)
- `compute-numthreads-functional` — `[numthreads(2,3,1)]` thread-ID + group-index linearization (CPU + CUDA + Vulkan)
- `compute-system-values-comprehensive` — `SV_DispatchThreadID` / `SV_GroupThreadID` / `SV_GroupID` / `SV_GroupIndex` cross-check on a 2×2×2 group (CPU + CUDA + Vulkan)
- `numthreads-3d-linearization` — verifies `SV_GroupIndex = x + y·Nx + z·Nx·Ny` for all 8 threads (CPU + CUDA + Vulkan)

**Groupshared memory & barriers** (4)
- `groupshared-barrier-functional` — write/barrier/read round-trip (CUDA + Vulkan) + HLSL emission
- `groupshared-barrier-spirv-emission` — SPIRV `OpControlBarrier` emission
- `groupshared-multi-barrier-functional` — two sequential write-barrier-read rounds chaining through groupshared (CUDA + Vulkan)
- `group-memory-barrier-no-sync-emission` — `GroupMemoryBarrier()` (no sync) emission across HLSL / GLSL / SPIRV

**Atomics** (2)
- `atomic-interlocked-functional` — concurrent `InterlockedAdd` under thread contention (Vulkan + CUDA)
- `atomic-interlocked-emission` — `InterlockedAdd` / `InterlockedMin` / `InterlockedCompareExchange` emission across HLSL / GLSL / SPIRV

**Wave ops — divergence & reconvergence** (10)
- `wave-divergent-if-functional` / `wave-divergent-if-emission` — divergent if/else, per-branch `WaveActiveMin`, reconverged `WaveActiveSum`
- `wave-switch-divergence-functional` — 4 switch cases, wave-sum after reconvergence
- `loop-divergence-reconvergence-functional` / `loop-divergence-reconvergence-emission` — threads iterate different counts, wave-sum after the loop
- `loop-break-wave-op-functional` — progressive attrition via `break`; `WaveActiveCountBits` decreases per iteration
- `early-return-wave-op-functional` — returned thread excluded from `WaveActiveMin`
- `nested-divergence-functional` — wave ops inside nested divergent blocks with progressive convergent-set shrinkage
- `wave-ballot-verify-functional` — `WaveActiveBallot` bit pattern + `WaveActiveCountBits` cross-check
- `multi-wave-thread-group-functional` — wave ops scoped to a single wave, not the full thread group

**Wave ops — base intrinsics** (3)
- `wave-active-all-any-functional` — `WaveActiveAllTrue` / `WaveActiveAnyTrue` (Vulkan + CUDA)
- `wave-intrinsics-emission` — `WaveIsFirstLane` / `WaveGetLaneIndex` / `WaveGetLaneCount` / `WaveActiveBallot` / `WaveReadLaneFirst` (HLSL + GLSL + SPIRV)
- `quad-ops-emission` — `QuadReadAcrossX` / `Y` / `Diagonal` and `QuadReadLaneAt` (HLSL + GLSL + SPIRV)

**Pipeline stages — entry points** (3)
- `multiple-entry-points-stages` — vertex, fragment, and compute entry points in one module
- `shader-attribute-stages` — `[shader("...")]` attribute on each stage
- `generic-entry-point` — non-generic entry point calling a generic helper, exercising implicit specialization (CPU + CUDA + Vulkan)

**Vertex / fragment** (5)
- `cross-stage-vertex-fragment-emission` — vertex→fragment data flow (HLSL)
- `vertex-system-values-emission` — `SV_VertexID` / `SV_InstanceID` / `SV_Position` (HLSL + GLSL + SPIRV)
- `vertex-return-struct-emission` — vertex returning a struct of varyings + `SV_Position` (SPIRV + GLSL)
- `fragment-system-values-emission` — `SV_Position` / `SV_IsFrontFace` / `SV_SampleIndex` / `SV_PrimitiveID` / `SV_Coverage` (HLSL + GLSL + SPIRV)
- `discard-fragment-emission` — HLSL `discard`, GLSL `discard`, SPIRV 1.6 `OpDemoteToHelperInvocation` (via `-target spirv-asm -emit-spirv-directly`)

**Fragment derivatives & helper threads** (1)
- `derivatives-helper-thread-emission` — `ddx` / `ddy` quad participation across HLSL / GLSL / SPIRV

**Varying-parameter legalization** (2)
- `half-varying-emission` — sub-word (`half`) vertex/fragment varyings (SPIRV)
- `varying-array-input` — array-typed vertex varying input, exercising the array-recursion path in `_createLegalVaryingVal` (HLSL + SPIRV)

**Specialty stages — emission** (4)
- `geometry-stream-emission` — `TriangleStream` + `Append` + `RestartStrip` (HLSL + GLSL + SPIRV)
- `tessellation-emission` — hull + domain shaders (HLSL + GLSL + SPIRV)
- `mesh-shader-emission` — mesh entry point (HLSL + SPIRV)
- `raytracing-stages-emission` — ray-generation / closest-hit / miss (HLSL DXIL + SPIRV)

Backends: functional tests use CPU + CUDA + Vulkan depending on feature
(GPU-less CI still covers CPU paths); emission tests target HLSL, GLSL,
and SPIRV.

## Test plan

- [x] `slang-test tests/diagnostics/execution-model/` passes locally (16 files)
- [x] `slang-test tests/language-feature/execution-model/` CPU + emission paths pass locally
- [x] CI: Vulkan and CUDA functional runs

### CODE REVIEW

File: tests/diagnostics/execution-model/unknown-stage-name-attribute.slang [hunk 1]
```diff
@@ -0,0 +1,11 @@
+// Test that using an unknown stage name in [shader(...)] attribute produces an error.
+
+//DIAGNOSTIC_TEST:SIMPLE(diag=CHECK):
+
+[shader("invalid")]
+/*CHECK:
+ ^^^^^^ unknown capability
+ ^^^^^^ unknown capability name 'invalid'.
+*/
```
[the remark this answers is not in the document]
Comment (jvepsalainen-nv) [reply tests/diagnostics/execution-model/unknown-stage-name-attribute.slang:9]: [Agent] Verified against actual compiler output: `slangc tests/diagnostics/execution-model/unknown-stage-name-attribute.slang` reports this diagnostic pointing at `shader` (column 2, 6 chars) for `[shader("invalid")]`, not at `invalid`. The CHECK carets match the compiler's actual span, which is what the test pins. Whether the compiler *should* point at the string literal is a separate question about compiler behavior.

File: tests/language-feature/execution-model/wavesize-emission.slang [hunk 2]
```diff
@@ -0,0 +1,11 @@
+// Test that WaveSize attribute is correctly emitted for HLSL target.
+
+//TEST:SIMPLE(filecheck=CHECK_HLSL): -target hlsl -stage compute -entry computeMain
+
+// CHECK_HLSL: [WaveSize(16)]
```
[the remark this answers is not in the document]
Comment (jvepsalainen-nv) [reply tests/language-feature/execution-model/wavesize-emission.slang]: [Agent] `wavesize-emission.slang` was removed in commit 93bdbc0 as redundant with existing `tests/hlsl/wave-size.slang`.

File: tests/diagnostics/execution-model/invalid-dispatch-thread-id-type.slang [hunk 3]
```diff
@@ -0,0 +1,12 @@
+// Test that SV_DispatchThreadID with a float type produces an error.
+
+//DIAGNOSTIC_TEST:SIMPLE(diag=CHECK):-target hlsl -stage compute -entry computeMain
+
+[numthreads(4, 1, 1)]
+void computeMain(float3 tid : SV_DispatchThreadID)
```
[the remark this answers is not in the document]
Comment (jvepsalainen-nv) [reply tests/diagnostics/execution-model/invalid-dispatch-thread-id-type.slang]: [Agent] The `uint4` boundary case was added as `computeMain2` in the same file (commit 93bdbc0).

File: tests/language-feature/execution-model/loop-break-wave-op-functional.slang [hunk 4]
```diff
@@ -0,0 +1,45 @@
...
+}
+
+// Thread 0: breaks at i=0, all 4 active → 4
+// Thread 1: breaks at i=1, threads {1,2,3} active → 3
+// Thread 2: breaks at i=2, threads {2,3} active → 2
+// Thread 3: breaks at i=3, thread {3} active → 1
+// CHECK: 4
+// CHECK: 3
+// CHECK: 2
+// CHECK: 1
```
[the remark this answers is not in the document]
Comment (jvepsalainen-nv) [reply tests/language-feature/execution-model/loop-break-wave-op-functional.slang:45]: [Agent] Acknowledged. Both `-vk` and `-cuda` run in CI; if either backend diverges from per-iteration active-mask semantics the test will catch it.

File: tests/language-feature/execution-model/multiple-entry-points-stages.slang [hunk 5]
```diff
@@ -0,0 +1,51 @@
...
+VsOut vertMain(float3 position : POSITION, float lum : LUMINANCE)
+{
+    VsOut o;
+    o.pos = float4(position, 1.0);
+    o.brightness = brighten(lum);
+    return o;
+}
+
+// --- Fragment entry point ---
+// CHECK_FRAG: SV_TARGET
```
[the remark this answers is not in the document]
Comment (jvepsalainen-nv) [reply tests/language-feature/execution-model/multiple-entry-points-stages.slang:36]: [Agent] Verified against actual HLSL emission: `slangc -target hlsl -entry fragMain -stage fragment` emits `SV_TARGET` (uppercase) for the return semantic. The `kIndexedSystemSemantics` table at `slang-emit-hlsl.cpp:2059` governs *indexed* semantics (e.g. `SV_Target0` with an explicit index), not the plain `SV_TARGET` written for an unindexed function-return semantic. The CHECK matches the actual output and the test passes.

File: tests/language-feature/execution-model/shader-attribute-stages.slang [hunk 6]
```diff
@@ -0,0 +1,42 @@
...
+[shader("vertex")]
+VsOut vsMain(float3 pos : POSITION)
+{
+    VsOut o;
+    o.pos = float4(pos, 1.0);
+    return o;
+}
+
+// --- Fragment shader ---
+// CHECK_FS: SV_TARGET
```
[the remark this answers is not in the document]
Comment (jvepsalainen-nv) [reply tests/language-feature/execution-model/shader-attribute-stages.slang:37]: [Agent] Verified against actual HLSL emission: `slangc -target hlsl -entry fsMain -stage fragment` emits `SV_TARGET` (uppercase) for the function-return semantic. The `kIndexedSystemSemantics` table at `slang-emit-hlsl.cpp:2059` governs indexed semantics only. The CHECK matches the actual output and the test passes.

File: tests/language-feature/execution-model/early-return-wave-op-functional.slang [hunk 7]
```diff
@@ -0,0 +1,36 @@
+// Corner case: early return removes threads, then remaining threads do wave ops.
+// Thread 0 returns early. Threads {1,2,3} continue and do WaveActiveMin.
+// WaveActiveMin({1,2,3}) should be 1 (thread 0 is inactive).
+//
+// Then thread 1 also returns. Threads {2,3} do WaveActiveSum.
+// WaveActiveSum({2,3}) should be 5.
```
[the remark this answers is not in the document]
Comment (jvepsalainen-nv) [reply tests/language-feature/execution-model/early-return-wave-op-functional.slang]: [Agent] The stale comment was removed in commit 93bdbc0; the current header (lines 1-3) matches the shader code.

File: tests/language-feature/execution-model/wave-intrinsics-emission.slang [hunk 8]
```diff
@@ -0,0 +1,45 @@
+// Test emission of core wave intrinsics across HLSL and GLSL targets.
+// From the docs: "Threads in a wave can synchronize and share data efficiently
+// using wave-tangled functions such as ballots, reductions, shuffling..."
+//
+// Verifies WaveIsFirstLane, WaveGetLaneIndex, WaveGetLaneCount,
+// WaveActiveBallot, and WaveReadLaneFirst emit correct target-specific code.
+
+//TEST:SIMPLE(filecheck=CHECK_HLSL): -target hlsl -stage compute -entry computeMain
+//TEST:SIMPLE(filecheck=CHECK_GLSL): -target glsl -stage compute -entry computeMain
```
[the remark this answers is not in the document]
Comment (jvepsalainen-nv) [reply tests/language-feature/execution-model/wave-intrinsics-emission.slang:9]: [Agent] SPIRV coverage is present in the current revision: `//TEST:SIMPLE(filecheck=CHECK_SPIRV): -target spirv-asm -stage compute -entry computeMain -emit-spirv-directly` at line 10, with `CHECK_SPIRV: OpGroupNonUniform*` lines for each intrinsic. Sibling `wave-divergent-if-emission.slang` and `loop-divergence-reconvergence-emission.slang` also have SPIRV checks.

File: tests/language-feature/execution-model/compute-numthreads-functional.slang [hunk 9]
```diff
@@ -0,0 +1,26 @@
+// Test that numthreads dimensions correctly map to thread IDs.
```
[the remark this answers is not in the document]
Comment (jvepsalainen-nv) [reply tests/language-feature/execution-model/compute-numthreads-functional.slang:1]: [Agent] The former `wavesize-emission.slang` was removed as redundant with `tests/hlsl/wave-size.slang` (commit 93bdbc0). The current PR description no longer references WaveSize emission — only the diagnostics test `invalid-wave-size` remains.

File: tests/diagnostics/execution-model/stage-dependency-incompatible.slang [hunk 10]
```diff
@@ -0,0 +1,24 @@
...
+//DIAGNOSTIC_TEST:SIMPLE(diag=CHECK1):-target spirv -entry computeMain
+//DIAGNOSTIC_TEST:SIMPLE(diag=CHECK2):-target hlsl -entry fragMain
+
+[shader("compute")]
+[numthreads(1, 1, 1)]
+void computeMain(out float depth : SV_Depth)
+//CHECK1:                  ^^^^^ system value semantic 'SV_Depth' cannot be used as output in 'compute' shader stage
+{
+    depth = 0.5;
+}
```
[the remark this answers is not in the document]
Comment (jvepsalainen-nv) [reply tests/diagnostics/execution-model/stage-dependency-incompatible.slang:14]: [Agent] Verified against actual compiler output: `slangc -target spirv -entry computeMain` reports this diagnostic at column 28 pointing at the parameter name `depth` (5 chars), not at `SV_Depth`. The CHECK1 caret matches the actual span. Whether the compiler *should* point at the semantic token instead is a separate question about compiler behavior.

File: tests/language-feature/execution-model/loop-divergence-reconvergence-emission.slang [hunk 11]
```diff
@@ -0,0 +1,41 @@
...
+// Verifies that wave ops emit correctly after a loop where threads iterate
+// different numbers of times (causing divergence), then reconverge.
+
+//TEST:SIMPLE(filecheck=CHECK_HLSL): -target hlsl -stage compute -entry computeMain
+//TEST:SIMPLE(filecheck=CHECK_GLSL): -target glsl -stage compute -entry computeMain
+//TEST:SIMPLE(filecheck=CHECK_SPIRV): -target spirv-asm -stage compute -entry computeMain -emit-spirv-directly
+
+RWStructuredBuffer<int> output;
+
+[numthreads(128, 1, 1)]
```
[the remark this answers is not in the document]
Comment (jvepsalainen-nv) [reply tests/language-feature/execution-model/loop-divergence-reconvergence-emission.slang:17]: [Agent] Applied in commit b68e2c09c. Added `CHECK_HLSL: [numthreads(128, 1, 1)]` and `CHECK_GLSL: layout(local_size_x = 128` for consistency with the sibling emission tests.

File: tests/diagnostics/execution-model/stage-incompatible-in-param.slang [hunk 12]
```diff
@@ -0,0 +1,9 @@
+//DIAGNOSTIC_TEST:SIMPLE(diag=CHECK):-target hlsl -entry main
+
+struct Payload { float3 data; };
+[shader("callable")]
+void main(in Payload data) {}
+//CHECK:             ^^^^ stage does not support in parameters
+//CHECK:             ^^^^ the 'callable' stage does not support `in` entry point parameters
+//CHECK:             ^^^^ stage does not support in parameters
+//CHECK:             ^^^^ the 'callable' stage does not support `in` entry point parameters
```
[the remark this answers is not in the document]
Comment (jvepsalainen-nv) [reply tests/diagnostics/execution-model/stage-incompatible-in-param.slang:13]: [Agent] Intentional. The compiler currently emits E39018 twice for the same invalid `in` parameter (once during entry-point semantic check, once during parameter binding). Added a comment in commit b68e2c09c documenting this so a future deduplication of the diagnostic is caught as an intentional test update.

File: tests/language-feature/execution-model/varying-array-input.slang [hunk 13]
```diff
@@ -0,0 +1,50 @@
+// Test that array-typed varying parameters legalize correctly.
+// Exercises the array-recursion path in `_createLegalVaryingVal`
+// (slang-ir-legalize-varying-params.cpp around L744-L828) which
+// handles the "outer array brackets" tracking when decomposing a
```
[the remark this answers is not in the document]
Comment (jvepsalainen-nv) [reply tests/language-feature/execution-model/varying-array-input.slang]: Fixed in `2e00868b4`. Dropped the stale line ranges from the three test comments (`varying-array-input.slang`, `vertex-return-struct-emission.slang`, `generic-entry-point.slang`); the function/symbol names already convey the intent.

File: tests/language-feature/execution-model/multiple-entry-points-stages.slang [hunk 14]
```diff
@@ -0,0 +1,54 @@
...
+    o.pos = float4(position, 1.0);
+    o.brightness = brighten(lum);
+    return o;
+}
+
+// --- Fragment entry point ---
+// CHECK_FRAG: brighten
+// CHECK_FRAG: SV_TARGET
+[shader("fragment")]
+float4 fragMain(float brightness : BRIGHTNESS) : SV_Target
```
[the remark this answers is not in the document]
Comment (jvepsalainen-nv) [reply tests/language-feature/execution-model/multiple-entry-points-stages.slang:38]: False positive — verified locally. Slang's HLSL emitter normalizes the SV_Target semantic to uppercase regardless of source casing; `./build/Debug/bin/slangc tests/.../multiple-entry-points-stages.slang -target hlsl -entry fragMain -stage fragment` emits `: SV_TARGET` (uppercase). Both this test and `shader-attribute-stages.slang` pass against the existing CHECK pattern.

File: tests/language-feature/execution-model/shader-attribute-stages.slang (hunk 6 above)
[the remark this answers is not in the document]
Comment (jvepsalainen-nv) [reply tests/language-feature/execution-model/shader-attribute-stages.slang:37]: False positive — same as the multiple-entry-points-stages thread. Slang's HLSL emitter uppercases SV_Target; the existing CHECK pattern matches the actual output and the test passes.

File: tests/diagnostics/execution-model/entry-point-no-stage.slang [hunk 15]
```diff
@@ -0,0 +1,7 @@
+//DIAGNOSTIC_TEST:SIMPLE(diag=CHECK):-target hlsl -entry main
```
[the remark this answers is not in the document]
Comment (jvepsalainen-nv) [reply tests/diagnostics/execution-model/entry-point-no-stage.slang]: Fixed in `2e00868b4`. Deleted the new duplicate; the existing `tests/diagnostics/entry-point-no-stage.slang` already pins E38007 with caret-aligned messages.

File: tests/diagnostics/execution-model/capability-stage-incompatible.slang [hunk 16]
```diff
@@ -0,0 +1,34 @@
+// Diagnostic: a function requiring `fragment` stage capability is
```
[the remark this answers is not in the document]
Comment (jvepsalainen-nv) [reply tests/diagnostics/execution-model/capability-stage-incompatible.slang:1]: [Agent] PR description has been rewritten to match the actual diff: 16 + 37 = 53 tests, with the swapped error codes for stage-incompatible-{in,out}-param corrected (E39018 for the in case, E39017+E41018 for the out case), the missing entry-point-no-stage entry removed, and capability-stage-incompatible / sv-output-direction-mismatch added under Stage / capability and System-value direction respectively.

**Review by expipiplus1 — APPROVED** [review]
